web3.nim

```diff
@@ -91,29 +95,61 @@ proc getHistoricalEvents(s: Subscription, options: JsonNode) {.async.} =
     let logs = await s.web3.provider.eth_getLogs(options)
     for l in logs:
       if s.removed: break
-      s.callback(l)
+      s.eventHandler(l)
     s.historicalEventsProcessed = true
     var i = 0
     while i < s.pendingEvents.len: # Mind reentrancy
       if s.removed: break
-      s.callback(s.pendingEvents[i])
+      s.eventHandler(s.pendingEvents[i])
       inc i
     s.pendingEvents = @[]
   except CatchableError as e:
     echo "Caught exception in getHistoricalEvents: ", e.msg
     echo e.getStackTrace()
 
-proc subscribe*(w: Web3, name: string, options: JsonNode, callback: proc(j: JsonNode) {.gcsafe.}): Future[Subscription] {.async.} =
-  var options = options
-  if options.isNil: options = newJNull()
-  let id = await w.provider.eth_subscribe(name, options)
-  result = Subscription(id: id, web3: w, callback: callback)
+proc subscribe*(w: Web3, name: string, options: JsonNode,
+                eventHandler: SubscriptionEventHandler,
```

Comment:
it tends to be hard to synchronize two handlers for these things - ie if it's a single subscription that gets events until it gets an error (or an error happens during an event), whoever is often needs to coordinate normal and error flow.. an easy way out is `Result[T, CatchableError]` + the special EH mode for result..

Reply:
...which also trivially allows building an adapter to use two handlers

Reply:
To be fair, building the said adapter is trivially possible in both directions.

I've considered something similar to what you suggest, but there were few arguments for the current approach:

1. Web3 is not a strict standard, but the practices of the libraries in the different languages are quite similar. The most frequently used web3 implementation is the JavaScript one and there you have these two handlers as separate callbacks:
https://web3js.readthedocs.io/en/v1.2.7/web3-eth-subscribe.html

      Some people may be interested in porting their JavaScript code to Nim and the current API makes it slightly easier to do so.

2) Error handling in the pub/sub pattern is a bit unusual. Many users will decide to ignore the errors. Having the error case in a separate callback allows you to reuse the same "onError" handler for multiple subscriptions (e.g. it could be a function that just prints a warning).

Nevertheless, I don't have strong objections to changing the API. I'm still evaluating what are the best ways to recover from errors in the interaction with the Web3 provider, so I expect that our experience will shape up the final API.

Reply:
not quite: turning one handler into two is trivially synchronous - one from two however requires synchronising the two information channels - usually a harder problem (you at least need to make assumptions that may break in the future, about the order of callbacks)